flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/AutoClosableProcess.java

```diff
@@ -117,14 +117,13 @@ public void runBlocking(final Duration timeout) throws IOException {
 
 		public void runBlockingWithRetry(final int maxRetries, final Duration attemptTimeout, final Duration globalTimeout) throws IOException {
 			int retries = 0;
-			Duration remainTimeout = globalTimeout;
+			long expectedTimeMillis = System.currentTimeMillis() + globalTimeout.toMillis();
```

Comment:
You can simplify the global timeout handling by converting it to a `Deadline` and using `#hasTimeLeft` as the exit condition:
```
Deadline deadline = Deadline.fromNow(globalTimeout);
while (true) {
	...
	} catch (Exception e) {
		if (++retries > maxRetries || (!deadline.hasTimeLeft)) {
			...
```

Reply:
Just saw the comment, thanks for the comment and appended commit, You are very nice. Thank you.